design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,264 @@
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  engineVersion: "9.6"
+```
+
+This model is powerful because it allows an application developer to create a resource claim without having to know the implementation details or even the underlying provider. However, the fact that every resource class is of the same `kind` presents a key issue: The required parameters for a resource class may vary widely, and they are currently only provided as an arbitrary map that is eventually read by the controller for the the specified `provisioner`. Therefore, an administrator who is creating resource classes does not know what fields are required and will not be notified of missing or extraneous fields until the provisioning of a resource that references the class.
+
+The `parameters` supplied by the resource class are used to populate the `spec` of the managed resource (i.e. the Kubernetes representation of the external resource) when it is created. For instance, the creation of `mysql-claim`, which references the `standard-mysql` class, is watched by the claim controller for AWS RDS instances. It brings together the information provided in the claim and class to create the `RDSInstance` managed resource. Specifically, it calls the `ConfigureMyRDSInstance()` function. As part of the configuration, the function creates the `spec` of the `RDSInstance` managed resource from the `parameters` of the `ResourceClass`:
+
+```go
+spec := v1alpha1.NewRDSInstanceSpec(cs.Parameters)
+```
+
+Looking at the `NewRDSInstanceSpec()` function reveals how the parameters are being parsed:
+
+```go
+// NewRDSInstanceSpec from properties map
+func NewRDSInstanceSpec(properties map[string]string) *RDSInstanceSpec {
+	spec := &RDSInstanceSpec{
+		ResourceSpec: corev1alpha1.ResourceSpec{
+			ReclaimPolicy: corev1alpha1.ReclaimRetain,
+		},
+	}
+
+	val, ok := properties["masterUsername"]
+	if ok {
+		spec.MasterUsername = val
+	}
+
+	val, ok = properties["engineVersion"]
+	if ok {
+		spec.EngineVersion = val
+	}
+
+	val, ok = properties["class"]
+	if ok {
+		spec.Class = val
+	}
+
+	val, ok = properties["size"]
+	if ok {
+		if size, err := strconv.Atoi(val); err == nil {
+			spec.Size = int64(size)
+		}
+	}
+
+	val, ok = properties["securityGroups"]
+	if ok {
+		spec.SecurityGroups = append(spec.SecurityGroups, strings.Split(val, ",")...)
+	}
+
+	val, ok = properties["subnetGroupName"]
+	if ok {
+		spec.SubnetGroupName = val
+	}
+
+	return spec
+}
+```
+
+This is a very rough parsing process and can result in missing parameters without notifying the Crossplane user.
+
+This problem can be solved by the implementation of "strongly typed" resource classes. Instead of every resource class being of `kind: ResourceClass`, they would map one-to-one with the concrete managed resource kind for which they provided specifications (i.e. `kind: RDSInstanceClass`).
+
+## Goals
+
+* Enable schema documentation and validation for resource classes.
```

Comment:
and annotations?

Reply:
This was not called out as an explicit goal because it is more of a byproduct of implementing strongly typed resource classes (i.e. the fact that strongly typed resource classes exist mean that different type's can be given different specific annotations). However, I don't have a problem with adding that as a goal here because it is a significant feature that is made available by these changes, and it would be useful to call that out in this document as motivation for moving to this pattern.